fix(chart-data): A lot of idle in transaction session when executing long chart data queries

Always-prog opened the pull request:
<!---
Please write the PR title following the conventions at https://www.conventionalcommits.org/en/v1.0.0/
Example:
fix(dashboard): load charts correctly
-->

### SUMMARY
In Superset's current implementation, when fetching data for charts, the metadata database (ex. - PostgreSQL) session remains open throughout the entire query execution. This leads to accumulation of numerous idle sessions in idle in transaction state, particularly under heavy user load or during long-running chart queries.

The common workarounds for this issue involve either is increasing PostgreSQL's max_connections to unreasonably high values (e.g., 1000+), or throttling workload through Celery task limits (when using Global Async Queries), which negatively impacts Superset's performance.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
Look at the count and  duration of "idle in transaction" sessions in DBeaver chart

# Before

https://github.com/user-attachments/assets/7c385785-e395-4283-bc46-db245203fb91


# After
https://github.com/user-attachments/assets/d2c46bc2-a928-4c14-a762-e3952fbbf610


In my PR I've modified the session handling logic:
 - Close the session before executing chart data query, releasing the metadata database connection
 - Reopen the session after query completion with loading previously removed SQLAlchemy objects 


### TESTING INSTRUCTIONS
1. Create a Session Monitoring Chart in DBeaver.
Use this SQL to track session states in your metadata database:
```sql
SELECT
  COUNT(*) FILTER (WHERE state = 'active') AS active_sessions,
  COUNT(*) FILTER (WHERE state = 'idle') AS idle_sessions,
  COUNT(*) FILTER (WHERE state = 'idle in transaction') AS idle_in_transaction
FROM pg_stat_activity
WHERE usename <> 'postgres' and datname = 'superset';
```
3. Simulate a Slow Query in Superset using next virtual dataset:
```sql
SELECT *
FROM generate_series(1, 50000000) a
CROSS JOIN generate_series(1, 50000000) b
LIMIT 1000000000;
```
4. Open this chart in Superset.
5. While it loads, observe the session metrics in DBeaver

<!--- Required! What steps can be taken to manually verify the changes? -->

### ADDITIONAL INFORMATION
<!--- Check any relevant boxes with "x" -->
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
- [x] Has associated issue: #27531
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

Always-prog committed:
added SuspendSession for chart data queries

rusackas commented on superset/utils/metadata.py:
```diff
@@ -0,0 +1,9 @@
+from superset import db
```
This should help CI pass ;)

villebro commented on superset/utils/metadata.py:
```diff
@@ -0,0 +1,9 @@
+        self.session_objects = db.session.identity_map.values()
+        db.session.remove()
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        for obj in self.session_objects:
+            db.session.add(obj)
```
I like the idea to release the metastore connection for the duration of the analytical query, but this seems slightly low level/dangerous. Have you considered any other approaches, like just doing `db.session.close()` and then letting the session lazily resume when needed?

michael-s-molina replied:
I agree with @villebro. Detaching/reattaching objects bring session sync problems. For example, you would need to use `merge` instead of `add` when reattaching given that the database might have changed while you wait for the analytical query to complete. `merge` would sync with the actual state of the database but also fire additional queries. Another thing is that `flush` would also need to be invoked before `remove` to avoid losing changes. In other words, there might be simpler alternatives.

Always-prog replied:
@michael-s-molina I'll try flush and merge, thank you.
The reason I manually add objects is to fix this problem
![image](https://github.com/user-attachments/assets/f9791622-9906-480c-a93c-e4b1a45c6cab)

Error raises because the relationships in loaded models, such as Sql a Table, Column, and SqlaMetric, with lazy=True (the default), are unloaded from memory after the session.remove() method is called. Accessing these objects after the session is closed will raise a DetachedInstanceError.


@villebro I was getting unexpected error using session.close(), but don't remember why. I'll try to use it again later.

Always-prog committed:
Update superset/utils/metadata.py

Co-authored-by: Evan Rusackas <evan@preset.io>

rusackas commented:
Re-ran CI, but it seems there are still some issues with pre-commit and such. Added a couple more reviewers for good measure.

mistercrunch commented:
Different solution to the same problem here https://github.com/apache/superset/pull/31315 . I like the approach in your PR if we can CI to go through. We should get a solution through for sure. I forgot why my PR was held up, I think we wanted to test the heck out of it but never got around to do so. Or maybe we didn't see exactly what we wanted to see over load testing I forgot..

Always-prog committed:
closing session instead of removing

Always-prog committed:
Merge branch 'fix/idle-in-session-during-query' of github.com:TechAudit-BI/apache_superset into fix/idle-in-session-during-query

Always-prog committed:
Merge branch 'master' into fix/idle-in-session-during-query

Always-prog committed:
- Using flush and merge
- Using contextmanager instead of with construction
- disable session suspend for SQL-lab

Always-prog committed:
trying to use add instead of merge

Always-prog committed:
return merge

Always-prog commented:
After hours of trying to pass all the test cases, I realized that the entire direction I was going in — trying to temporarily close the session and "bring everything back" — is flawed and won’t work.

---

### What I tried and what I discovered

I attempted to temporarily close the SQLAlchemy session during queries to datasources to avoid idle-in-transaction sessions. My approach was:

1. Save all objects from the current session via `identity_map`.
2. Call `db.session.close()` to release the connection.
3. After the task finishes, reopen the session and call `db.session.merge()` on each object to re-attach them.

This worked in simple, isolated cases — such as when loading chart data — where no new objects were created and no state was reused across the session boundary.

However, I ran into serious issues in datasource-related test cases and in SQL Lab, especially around the `Query` model.

---

### A representative issue with the `Query` object

In SQL Lab, a `Query` instance is created and added to the session as seen here:
[https://github.com/apache/superset/blob/e2a22d481c0560eb3afb317367765b62dd46c310/superset/commands/sql\_lab/execute.py#L144](https://github.com/apache/superset/blob/e2a22d481c0560eb3afb317367765b62dd46c310/superset/commands/sql_lab/execute.py#L144)

When I inspect the identity map before closing the session:

```python
logger.info(list(db.session.identity_map.values()))
# Output: [..., <Query at 0x7f14453c7310>, ...]
```

This shows the `Query` object in memory at `0x7f14453c7310`.

But once I call `db.session.close()`, the instance becomes detached — effectively lost from SQLAlchemy's tracking system.

Later, when I call:

```python
db.session.merge(query)
```
…it creates a **new** `Query` instance with a different memory address. Any existing references to the old object now point to a stale, invalid object. Trying to use that variable (`query`) in the code leads to errors like:

> `Instance '<Query at 0x7f14453c7310>' is not persistent within this Session.`
> `Object was deleted or detached.`

---

### Why it fails

This breaks because **SQLAlchemy sessions are not designed to be paused and resumed**.

Once you close a session:

* All attached objects become detached.
* They cannot be used unless re-merged.
* Any code still holding a reference to an old instance is now working with an invalid object.

Even calling `flush()` before closing the session doesn’t help. It pushes changes to the DB, but the objects are still bound to the session.

And this problem doesn’t just affect the `Query` model — it will happen with **any SQLAlchemy model**, whether it exists today or is added in the future.

---

### Conclusion

The key takeaway:

> **Closing a session and trying to "restore everything as it was" is fundamentally incompatible with how SQLAlchemy works.**

This kind of logic might work in small, single-purpose tasks, but it breaks completely when:

* new objects are created after the session starts,
* objects are reused across the codebase,
* or internal state depends on session identity.

This leads to detached references, subtle bugs, and hard-to-diagnose failures.

---

### So what *can* be done to fix idle-in-transaction issue?

The proper long-term solution is to **refactor the code around fetch data from datasources** — making it safe to:
* close the session without losing any critical objects,
* reopen it later, and
* *explicitly re-fetch all required objects from the database* (instead of trying to preserve state in memory).

But that’s a significant amount of work — and not something that can be hacked around with `merge()` and `identity_map`.

mistercrunch commented:
As I mentioned I spent some cycles hacking a fix for it here too https://github.com/apache/superset/pull/31315, I believe the solution works but need thorough tested. It's not clean but I believe it releases connections. It's not elegant but may work.

Few related points that were not touched here:
- if you use connection pools, I think this is less of a problem, or at least solutions can be designed differently. At Preset we have our own ways of dealing with this as it relates to our multitenancy-specific implementation. Regardless the solution/fix here ideally need to work with different connection pool configurations. There's a bucket of complexity around greenlet/gevent stuff here, and the fact that some database drivers are not thread-safe
- the real solution is to route all analytical workloads through "async queries" / celery workers. Ideally flask only serves normal, short lived web requests, flask should never wait on some external database we don't control and pass workloads over to celery. This solves for cloggage issues in web servers, but doesn't solve for "releasing the metadata database connections during extended external workloads". Logic need to be redefined for that regardless of where the work takes place...

mistercrunch commented:
About the "data access layer", clearly it's SIP-worthy, but don't want to publish more SIP than I can commit to push through personally or sponsor...

Always-prog committed:
Merge branch 'master' into fix/idle-in-session-during-query

rusackas commented:
Do we want to keep this PR going and tune things up here, or close it in hopes of finding another approach and/or opening a SIP?

mistercrunch commented:
Did some more work on https://github.com/apache/superset/pull/31315, looking for people to help at testing it in different settings.
